altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright (C) 2017  AltOS-Rust Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//! Linked list code for the the memory allocator
+//! This is intended for use by the free_list_allocator
+//!
+
+use core::{mem, ptr};
+
+use alignment;
+
+// BlockHeader keeps track of a free block of memory
+#[cfg(not(test))]
+#[repr(C)]
+pub struct BlockHeader {
+    block_size: usize,
+    next_block: *mut BlockHeader,
+}
+
+impl BlockHeader {
+    const fn new(size: usize) -> Self {
+        BlockHeader {
+            block_size: size,
+            next_block: ptr::null_mut(),
+        }
+    }
+}
+
+// FreeList is a linked list which keeps track of free blocks of memory
+// Free blocks are embedded in the free memory itself
+#[cfg(not(test))]
+pub struct FreeList {
+    block_hdr_size: usize,
+    head: *mut BlockHeader,
+}
+
+// These are (trivially) implemented so FreeList objects can be passed between threads.
+unsafe impl Send for FreeList {}
+unsafe impl Sync for FreeList {}
+
+impl FreeList {
+    pub const fn new() -> Self {
+        FreeList {
+            block_hdr_size: 0,
+            head: ptr::null_mut(),
+        }
+    }
+
+    pub fn init(&mut self, heap_start: usize, heap_size: usize) {
+        self.block_hdr_size = mem::size_of::<BlockHeader>();
+        // Forcing BlockHeader size to be a power of two avoids weird issues with alignment
+        // and extra logic we would have to include elsewhere
+        if !self.block_hdr_size.is_power_of_two() {
+            panic!("block header size is not power of two");
+        }
```

Comment:
We were originally planning on doing a doubly linked list. I put this in because I figured if we made a change that made the block_hdr_size _not_ a power of two, that could cause problems, so we'd want to check for it. We only know it's going to be a power of two because it's 2 usize fields, but if we changed this to be a doubly linked list it would 3 fields, resulting in 12 bytes. I guess it probably doesn't make sense to leave it in though.